tests/test_client.py

```diff
@@ -0,0 +1,151 @@
+             "n": "ZgKgqvEo_GZMamwy293IvA",
+             "alg": "RS256"}]},
+        None
+    )
+    with mock.patch("portier.client.jwt") as mocked_jwt:
+        mocked_jwt.decode.return_value = {
+            "sub": "invalid subject"
+        }
+        with pytest.raises(ValueError) as e:
+            get_verified_email("http://broker-url.tld/",
+                               "eyJraWQiOiAiYWJjIn0.foo.bar",
+                               "audience", "issuer", cache)
+        assert "Invalid email address: invalid subject" in str(e)
+
+
+def test_get_verified_email_validate_it_can_find_a_public_key():
+    cache = mock.MagicMock()
+    cache.get.side_effect = (
+        {"keys": []},
+        None
+    )
+    with mock.patch("portier.client.jwt") as mocked_jwt:
+        mocked_jwt.decode.return_value = {
+            "sub": "invalid subject"
+        }
+        with pytest.raises(ValueError) as e:
+            get_verified_email("http://broker-url.tld/",
+                               "eyJraWQiOiAiYWJjIn0.foo.bar",
+                               "audience", "issuer", cache)
+        assert "Cannot find public key with ID abc" in str(e)
+
+
+def test_get_verified_email_validate_it_can_decode_the_jwt_payload():
+    cache = mock.MagicMock()
+    cache.get.side_effect = (
+        {"keys": [
+            {"kid": "abc",
+             "e": "KHTPnNouCvwROWeIWQkJiw",
+             "n": "ZgKgqvEo_GZMamwy293IvA",
+             "alg": "RS256"}]},
+        None
+    )
+    with mock.patch("portier.client.jwt") as mocked_jwt:
+        mocked_jwt.decode.side_effect = Exception("Foobar")
+        with pytest.raises(ValueError) as e:
+            get_verified_email("http://broker-url.tld/",
+                               "eyJraWQiOiAiYWJjIn0.foo.bar",
+                               "audience", "issuer", cache)
+        assert "Invalid JWT: Foobar" in str(e)
+
+
+def test_get_verified_email_validate_the_nonce():
+    cache = mock.MagicMock()
+    cache.get.side_effect = (
+        {"keys": [
+            {"kid": "abc",
+             "e": "KHTPnNouCvwROWeIWQkJiw",
+             "n": "ZgKgqvEo_GZMamwy293IvA",
+             "alg": "RS256"}]},
+        None
+    )
+    with mock.patch("portier.client.jwt") as mocked_jwt:
+        mocked_jwt.decode.return_value = {
+            "sub": "foobar@restmail.com",
+            "nonce": "a nonce"
+        }
+        with pytest.raises(ValueError) as e:
+            get_verified_email("http://broker-url.tld/",
+                               "eyJraWQiOiAiYWJjIn0.foo.bar",
+                               "audience", "issuer", cache)
+        assert "Invalid, expired, or re-used nonce" in str(e)
+
+
+def test_get_verified_return_the_subject_and_redirect_uri():
+    cache = mock.MagicMock()
+    cache.get.side_effect = (
+        {"keys": [
+            {"kid": "abc",
+             "e": "KHTPnNouCvwROWeIWQkJiw",
+             "n": "ZgKgqvEo_GZMamwy293IvA",
+             "alg": "RS256"}]},
```

Comment:
I have the feeling that tests could be a lot easier to read if those crazy values would be declared once on top (eg. `warm_cache`, `TOKEN`, etc...)